File: architecture-skills/design-patterns/examples/iterator/iterator.php
<?php
// Iterator pattern in PHP.
//
// A binary search tree traversed in sorted order. PHP builds the
// pattern in: implementing IteratorAggregate with a generator (yield)
// plugs a custom collection into foreach and iterator_to_array — no
// hand-rolled current/next/valid Iterator class needed.

declare(strict_types=1);

final class Node
{
    public ?Node $left = null;
    public ?Node $right = null;

    public function __construct(public readonly int $value)
    {
    }
}

/** @implements IteratorAggregate<int, int> */
final class SortedTree implements IteratorAggregate
{
    private ?Node $root = null;

    public function add(int $value): void
    {
        if ($this->root === null) {
            $this->root = new Node($value);
            return;
        }
        $node = $this->root;
        while (true) {
            if ($value < $node->value) {
                if ($node->left === null) {
                    $node->left = new Node($value);
                    return;
                }
                $node = $node->left;
            } else {
                if ($node->right === null) {
                    $node->right = new Node($value);
                    return;
                }
                $node = $node->right;
            }
        }
    }

    public function getIterator(): Generator
    {
        yield from $this->walk($this->root);
    }

    private function walk(?Node $node): Generator
    {
        if ($node !== null) {
            yield from $this->walk($node->left);
            yield $node->value;
            yield from $this->walk($node->right);
        }
    }
}

$tree = new SortedTree();
foreach ([5, 3, 8, 1, 4, 9, 7] as $value) {
    $tree->add($value);
}

$inOrder = iterator_to_array($tree, false);
echo 'in order: ' . implode(' ', $inOrder) . "\n";
echo 'sum: ' . array_sum($inOrder) . "\n";

foreach ($tree as $value) {
    if ($value > 4) {
        echo "first value above 4: $value\n";
        break; // generators support early exit cleanly
    }
}

File: architecture-skills/design-patterns/examples/iterator/iterator.cs
// Iterator pattern in C#.
//
// A binary search tree traversed in sorted order. C# builds the pattern
// in as IEnumerable<T>: an iterator method with yield return generates
// the iterator object, so no hand-rolled MoveNext/Current class is
// needed — and LINQ works on the collection for free.

using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;

public sealed class SortedTree : IEnumerable<int>
{
    private sealed class Node
    {
        public int Value;
        public Node? Left, Right;
    }

    private Node? _root;

    public void Add(int value)
    {
        if (_root is null)
        {
            _root = new Node { Value = value };
            return;
        }
        var node = _root;
        while (true)
        {
            if (value < node.Value)
            {
                if (node.Left is null) { node.Left = new Node { Value = value }; return; }
                node = node.Left;
            }
            else
            {
                if (node.Right is null) { node.Right = new Node { Value = value }; return; }
                node = node.Right;
            }
        }
    }

    public IEnumerator<int> GetEnumerator() => Walk(_root).GetEnumerator();

    IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();

    private static IEnumerable<int> Walk(Node? node)
    {
        if (node is null)
        {
            yield break;
        }
        foreach (int value in Walk(node.Left)) yield return value;
        yield return node.Value;
        foreach (int value in Walk(node.Right)) yield return value;
    }
}

public static class Program
{
    public static void Main()
    {
        var tree = new SortedTree();
        foreach (int value in new[] { 5, 3, 8, 1, 4, 9, 7 })
        {
            tree.Add(value);
        }

        Console.WriteLine("in order: " + string.Join(" ", tree));
        Console.WriteLine("sum: " + tree.Sum());
        Console.WriteLine("evens: " + string.Join(" ", tree.Where(v => v % 2 == 0)));
    }
}
